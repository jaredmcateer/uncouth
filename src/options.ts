import ts from "typescript";
import { ASTConvertPlugins } from "./plugins/types";
import { getDefaultPlugins } from "./plugins";
import * as vueTemplateCompiler from "vue-template-compiler";

export interface UncouthOptions {
  root: string;
  debug: boolean;
  compatible: boolean;
  setupPropsKey: string;
  setupContextKey: string;
  typescript: typeof ts;
  vueTemplateCompiler: typeof vueTemplateCompiler;
  prettierConfig: string;
  plugins: ASTConvertPlugins;
}

export type InputUncouthOptions = Partial<UncouthOptions>;

export function getDefaultUncouthOptions(tsModule: typeof ts = ts): UncouthOptions {
  return {
    root: process.cwd(),
    debug: false,
    compatible: false,
    setupPropsKey: "props",
    setupContextKey: "context",
    typescript: tsModule,
    vueTemplateCompiler: vueTemplateCompiler,
    prettierConfig: ".prettierrc",
    plugins: getDefaultPlugins(tsModule),
  };
}

export function mergeUncouthOptions(
  original: UncouthOptions,
  merged: InputUncouthOptions
): UncouthOptions {
  return {
    ...original,
    ...merged,
  };
}
