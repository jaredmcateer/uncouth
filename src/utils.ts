import type vueTemplateParser from "vue-template-compiler";
import type ts from "typescript";
import { ASTResult, ASTResultKind, ReferenceKind } from "./plugins/types";
import { TsHelper } from "./helpers/TsHelper";

export function isVueFile(path: string): boolean {
  return path.endsWith(".vue");
}
export function parseVueFile(
  vueTemplateParserModule: typeof vueTemplateParser,
  content: string
): vueTemplateParser.SFCDescriptor {
  return vueTemplateParserModule.parseComponent(content);
}

export function getNodeFromExportNode(
  tsModule: typeof ts,
  exportExpr: ts.Node
): ts.ClassDeclaration | undefined {
  switch (exportExpr.kind) {
    case tsModule.SyntaxKind.ClassDeclaration:
      return exportExpr as ts.ClassDeclaration;
  }
  return undefined;
}

export function getDefaultExportNode(
  tsModule: typeof ts,
  sourceFile: ts.SourceFile
): ts.ClassDeclaration | undefined {
  const exportStmts = sourceFile.statements.filter(
    (st) => st.kind === tsModule.SyntaxKind.ClassDeclaration
  );
  if (exportStmts.length === 0) {
    return undefined;
  }
  const exportNode = exportStmts[0] as ts.ClassDeclaration;

  return getNodeFromExportNode(tsModule, exportNode);
}

export function getDecoratorNames(tsModule: typeof ts, node: ts.Node): string[] {
  if (node.decorators) {
    return node.decorators.map((el) => {
      if (tsModule.isCallExpression(el.expression)) {
        return el.expression.expression.getText();
      } else {
        return el.expression.getText();
      }
    });
  }

  return [];
}

const $internalHooks = new Map<string, string | false>([
  ["beforeCreate", false],
  ["created", false],
  ["beforeMount", "onBeforeMount"],
  ["mounted", "onMounted"],
  ["beforeDestroy", "onBeforeUnmount"],
  ["destroyed", "onUnmounted"],
  ["beforeUpdate", "onBeforeUpdate"],
  ["updated", "onUpdated"],
  ["activated", "onActivated"],
  ["deactivated", "onDeactivated"],
  ["render", "onRender"],
  ["errorCaptured", "onErrorCaptured"], // 2.5
  ["serverPrefetch", "onServerPrefetch"], // 2.6
]);

export function isInternalHook(methodName: string): boolean {
  return $internalHooks.has(methodName);
}

export function getMappedHook(methodName: string): string | undefined | false {
  return $internalHooks.get(methodName);
}

export function isPrimitiveType(tsModule: typeof ts, returnType: ts.Type): boolean {
  return (
    !!(returnType.flags & tsModule.TypeFlags.NumberLike) ||
    !!(returnType.flags & tsModule.TypeFlags.StringLike) ||
    !!(returnType.flags & tsModule.TypeFlags.BooleanLike) ||
    !!(returnType.flags & tsModule.TypeFlags.Null) ||
    !!(returnType.flags & tsModule.TypeFlags.Undefined)
  );
}

export function convertNodeToASTResult<T extends ts.Node>(
  tsHelper: TsHelper,
  node: T
): ASTResult<T> {
  return {
    imports: [],
    kind: ASTResultKind.OBJECT,
    reference: ReferenceKind.NONE,
    attributes: [],
    tag: "InheritObjProperty",
    nodes: [tsHelper.addTodoComment(node, "Can't convert this object property.", false)],
  };
}

export function isString(val: unknown): val is string {
  return typeof val === "string";
}

export function isValidIdentifier(identifier: string): boolean {
  return /^[_$a-zA-Z\xA0-\uFFFF][_$a-zA-Z0-9\xA0-\uFFFF]*$/.test(identifier);
}
