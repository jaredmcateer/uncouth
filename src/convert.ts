import type ts from "typescript";
import { getDecoratorNames, getDefaultExportNode } from "./utils";
import { runPlugins } from "./plugins";
import { UncouthOptions } from "./options";
import { log } from "./debug";

const vueClassModules = ["vue-class-component", "vue-property-decorator"];

export function convertAST(
  sourceFile: ts.SourceFile,
  options: UncouthOptions,
  program: ts.Program
): string {
  const tsModule = options.typescript;

  log("check vue class library");
  const vueClassModuleImportStatement = sourceFile.statements.find((statement) => {
    if (tsModule.isImportDeclaration(statement)) {
      if (vueClassModules.includes((statement.moduleSpecifier as ts.StringLiteral).text)) {
        return true;
      }
    }
    return false;
  });
  if (!vueClassModuleImportStatement) {
    throw new Error("no vue class library in this file.");
  }

  log("check default export class");
  const defaultExportNode = getDefaultExportNode(options.typescript, sourceFile);
  if (!defaultExportNode) {
    throw new Error("no default export class");
  }

  const otherStatements = sourceFile.statements
    .map((el) => el)
    .filter(
      (el) =>
        !(
          (tsModule.isClassDeclaration(el) &&
            getDecoratorNames(tsModule, el).includes("Component")) ||
          (tsModule.isImportDeclaration(el) &&
            vueClassModules.includes((el.moduleSpecifier as ts.StringLiteral).text)) ||
          (tsModule.isImportDeclaration(el) &&
            (el.moduleSpecifier as ts.StringLiteral).text === "vue")
        )
    );

  let resultStatements = [...otherStatements, ...runPlugins(defaultExportNode, options, program)];

  resultStatements = [
    ...resultStatements.filter((el) => tsModule.isImportDeclaration(el)),
    ...resultStatements.filter((el) => !tsModule.isImportDeclaration(el)),
  ];

  log("output result code");
  const printer = tsModule.createPrinter();
  const { factory } = tsModule;
  const result = printer.printFile(factory.updateSourceFile(sourceFile, resultStatements));

  return result;
}
