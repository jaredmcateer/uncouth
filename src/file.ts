import path from "path";
import fs from "fs";
import { isVueFile, parseVueFile } from "./utils";
import { UncouthOptions } from "./options";
import { log } from "./debug";

export enum FileKind {
  VUE,
  TS,
}

export type FileInfo =
  | {
      fsPath: string;
      kind: FileKind;
      content: string;
    }
  | {
      fsPath: string;
      kind: FileKind;
      fileContent: string;
      start: number;
      end: number;
      content: string;
    };

export function readVueSFCOrTsFile(filePath: string, options: UncouthOptions): FileInfo {
  const fileFsPath = path.isAbsolute(filePath) ? filePath : path.resolve(options.root, filePath);
  const fileContent = fs.readFileSync(fileFsPath, { encoding: "utf8" });
  if (isVueFile(fileFsPath)) {
    const scriptContent = parseVueFile(options.vueTemplateCompiler, fileContent).script;
    if (scriptContent) {
      log(`Readed Vue file: ${fileFsPath}`);
      return {
        fsPath: fileFsPath,
        kind: FileKind.VUE,
        start: scriptContent.start,
        end: scriptContent.end,
        content: scriptContent.content,
        fileContent,
      };
    }
    throw new Error("The Vue SFC don't have sciprt element.");
  } else {
    log(`Readed TS file: ${fileFsPath}`);
    return {
      fsPath: fileFsPath,
      kind: FileKind.TS,
      content: fileContent,
    };
  }
}

export function writeFileInfo(fileInfo: FileInfo, content: string): void {
  if ("start" in fileInfo) {
    log(`Write Vue file: ${fileInfo.fsPath}`);
    const fileContent = `${fileInfo.fileContent.slice(
      0,
      fileInfo.start
    )}\n${content}${fileInfo.fileContent.slice(fileInfo.end)}`;
    fs.writeFileSync(fileInfo.fsPath, fileContent, { encoding: "utf8" });
  } else {
    log(`Write TS file: ${fileInfo.fsPath}`);
    fs.writeFileSync(fileInfo.fsPath, content, { encoding: "utf8" });
  }
}

export function existsFileSync(path: string): boolean {
  return fs.existsSync(path);
}
