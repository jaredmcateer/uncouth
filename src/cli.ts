import program from "commander";
import { convertFile } from "./index.js";
import inquirer from "inquirer";
import { writeFileInfo } from "./file";
import { version } from "../package.json";

function camelize(str: string) {
  return str.replace(/-(\w)/g, (_, c: string) => (c ? c.toUpperCase() : ""));
}

function getCmdOptions(cmd: { options: Array<{ long: string }> }) {
  const args: { [key: string]: boolean | string } = {};
  cmd.options.forEach((o: { long: string }) => {
    const key = camelize(o.long.replace(/^--/, ""));

    if (
      typeof (cmd as unknown as Record<string, string>)[key] !== "function" &&
      typeof (cmd as unknown as Record<string, string>)[key] !== "undefined"
    ) {
      args[key] = (cmd as unknown as Record<string, string>)[key];
    }
  });
  return args;
}

program.version(version).usage("<command> [options]");

program
  .command("single <filePath>")
  .description("convert vue component file from class to composition api")
  .option("-v, --view", "Output file content on stdout, and no write file.")
  .option("-o, --output", "Output result file path.")
  .option("-r, --root <root>", "Set root path for calc file absolute path. Default:`process.cwd()`")
  .option("-c, --config <config>", "Set uncouth config file path. Default: `'.uncouth.js'`")
  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  .action(async (filePath: string, cmd) => {
    const cmdOptions = getCmdOptions(cmd);
    if (!cmdOptions.output && !cmdOptions.view) {
      const result = await inquirer.prompt({
        name: "ok",
        type: "confirm",
        message:
          "You aren't using -o option to set output file path, It will replace original file content.",
      });
      if (!result.ok) {
        return;
      }
    }

    const { file, result } = convertFile(
      filePath,
      cmdOptions.root as string,
      cmdOptions.config as string
    );
    if (cmdOptions.view) {
      console.log(result);
      return;
    }

    writeFileInfo(file, result);
    console.log("Please check the TODO comments on result.");
  });

program.parse(process.argv);
