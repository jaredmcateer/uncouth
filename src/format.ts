/* eslint-disable @typescript-eslint/no-var-requires */
import { UncouthOptions } from "./options";
import path from "path";
import { log } from "./debug";
import prettier from "prettier/standalone";
import prettierTypescriptParser from "prettier/parser-typescript";
import { existsFileSync } from "./file";

export function format(content: string, options: UncouthOptions): string {
  const isNode = typeof window === "undefined";
  if (!isNode) {
    return prettier.format(content, {
      plugins: [prettierTypescriptParser],
      parser: "typescript",
      semi: false,
      singleQuote: true,
    });
  }

  const eslintConfigPath = path.resolve(options.root, options.eslintConfigFile);
  const prettierFormat = require("prettier-eslint") as (config: unknown) => string;
  const prettierEslintOpions = existsFileSync(eslintConfigPath)
    ? {
        text: content,
        filePath: eslintConfigPath,
        prettierOptions: {
          parser: "typescript",
        },
        fallbackPrettierOptions: {
          parser: "typescript",
        },
      }
    : {
        text: content,
        filePath: "",
        eslintConfig: {
          parser: require.resolve("@typescript-eslint/parser"),
          parserOptions: {
            sourceType: "module",
            ecmaFeatures: {
              jsx: false,
            },
          },
          rules: {
            semi: ["error", "never"],
            "padding-line-between-statements": [
              "error",
              { blankLine: "always", prev: "*", next: "export" },
              { blankLine: "always", prev: "const", next: "*" },
              { blankLine: "always", prev: "*", next: "const" },
            ],
          },
        },
        prettierOptions: {
          parser: "typescript",
          Semicolons: false,
          singleQuote: true,
          trailingComma: "none",
        },
        fallbackPrettierOptions: {
          parser: "typescript",
          singleQuote: true,
          Semicolons: false,
          trailingComma: "none",
        },
      };

  log("Format result code.....");
  return prettierFormat(prettierEslintOpions);
}
