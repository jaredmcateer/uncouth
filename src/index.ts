import { getSingleFileProgram } from './parser'
import { convertAST } from './convert'
import { InputUncouthOptions, getDefaultUncouthOptions, mergeUncouthOptions } from './options'
import { format } from './format'
import path from 'path'
import { readVueSFCOrTsFile, existsFileSync, FileInfo } from './file'
import { setDebugMode } from './debug'
import * as BuiltInPlugins from './plugins/builtIn'

export function convert (content: string, inputOptions: InputUncouthOptions): string {
  const options = mergeUncouthOptions(getDefaultUncouthOptions(inputOptions.typescript), inputOptions)
  const { ast, program } = getSingleFileProgram(content, options)

  return format(convertAST(ast, options, program), options)
}

export function convertFile (filePath: string, root: string, config: string): { file: FileInfo, result: string } {
  root = (typeof root === 'string')
    ? (
      path.isAbsolute(root) ? root : path.resolve(process.cwd(), root)
    )
    : process.cwd()
  config = (typeof config === 'string') ? config : '.uncouth.js'
  if (config.endsWith('.ts')) {
    require('ts-node/register')
  }
  const inputOptions: InputUncouthOptions = existsFileSync(path.resolve(root, config))
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    ? require(path.resolve(root, config)) as InputUncouthOptions
    : {}
  const options = mergeUncouthOptions(getDefaultUncouthOptions(inputOptions.typescript), inputOptions)
  options.root = root

  if (options.debug) {
    setDebugMode(true)
  }

  const file = readVueSFCOrTsFile(filePath, options)
  return {
    file,
    result: convert(file.content, options)
  }
}

export * from './plugins/types'
export { BuiltInPlugins }
export * from './utils'
export { getDefaultUncouthOptions, UncouthOptions } from './options'
