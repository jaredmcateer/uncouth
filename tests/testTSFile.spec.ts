import { convertFile } from "../src";
import { FileKind } from "../src/file";
import path from "path";
import util from "util";
import { exec } from "child_process";

const execAsync = util.promisify(exec);

describe("testTSFile", () => {
  const filePath = "fixture/Input.ts";

  it("compatible", () => {
    const { file, result } = convertFile(filePath, __dirname, "config/.compatible.uncouth.js");
    expect(file.fsPath.includes(path.basename(filePath))).toBeTruthy();
    expect(path.isAbsolute(file.fsPath)).toBeTruthy();
    expect(file.kind).toBe(FileKind.TS);
    expect(file).not.toHaveProperty("start");
    expect(file).not.toHaveProperty("end");
    expect(result).toMatchSnapshot();
  });

  it("no compatible", () => {
    const { file, result } = convertFile(filePath, __dirname, "config/.nocompatible.uncouth.js");
    expect(file.fsPath.includes(path.basename(filePath))).toBeTruthy();
    expect(path.isAbsolute(file.fsPath)).toBeTruthy();
    expect(file.kind).toBe(FileKind.TS);
    expect(file).not.toHaveProperty("start");
    expect(file).not.toHaveProperty("end");
    expect(result).toMatchSnapshot();
  });

  it("compatible and ts config file", async (done) => {
    jest.setTimeout(10000);
    const { stdout, stderr } = await execAsync(
      `node ${path.resolve(
        __dirname,
        "../bin/uncouth"
      )} single -v -r ${__dirname} -c config/.compatible.uncouth.ts ${filePath}`
    );

    expect(stdout).toMatchSnapshot("stdout");
    expect(stderr).toMatchSnapshot("stderr");

    done();
  });
});
