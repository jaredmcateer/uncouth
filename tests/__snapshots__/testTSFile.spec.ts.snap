// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testTSFile compatible 1`] = `
"import {
  computed,
  PropType,
  reactive,
  toRefs,
  ref,
  onBeforeUnmount,
  onUnmounted,
  watch,
  onMounted,
  defineComponent,
} from \\"@vue/composition-api\\";
import { useStore } from \\"vuex\\";
/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: \\"checked\\",
    event: \\"change\\",
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Boolean, default: false },
    bar: { type: Number as PropType<number>, default: 1 },
    foobar: { type: Object as PropType<CustomType> },
    noType: {},
    stringType: { type: String },
    booleanType: { type: Boolean },
    numberType: { type: Number },
    stringLiteralType: { type: String as PropType<\\"foo\\"> },
    numericLiteral: { type: Number as PropType<2> },
    booleanLiteral: { type: Boolean as PropType<true> },
    /* TODO: Too complex to determine primitive type: true | \\"false\\" */ complexType:
      {},
    /* TODO: Too complex to determine primitive type: string[] */ complexType2:
      {},
  },
  name: \\"oao\\",
  setup(props, context) {
    const store = useStore();
    /**
     * My count
     */
    const count = computed(() => {
      return /* TODO: Check this convert result, it can work well in 80% of cases.*/ context
        .root.$store.state.count;
    });
    const { a } = (() => {
      const a = \\"pa\\";
      return toRefs(
        reactive({
          a: a,
        })
      );
    })();
    const anotherComponent = ref<HTMLElement>(null);
    /* TODO: check function dispatch call signature is correct*/ const actA =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"actA\\", arg);
      };
    /* TODO: check function dispatch call signature is correct*/ const actD =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"namespace/actD\\", arg);
      };
    const actE = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"namespace/actE\\", val);
    };
    const actuhGee = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(actG, val);
    };
    const actH = async (): Promise<number> => {
      return await store.dispatch(\\"actH\\");
    };
    /* TODO: check function dispatch call signature is correct*/ const mutateA =
      (arg: unknown): unknown => {
        return store.commit(\\"mutateA\\", arg);
      };
    /* TODO: check function dispatch call signature is correct*/ const mutateB =
      (arg: unknown): unknown => {
        return store.commit(\\"namespace/mutateB\\", arg);
      };
    const mutateE = (val: \\"foo\\" | \\"bar\\"): number => {
      return store.commit(\\"namespace/mutateE\\", val);
    };
    const mGee = (val: \\"foo\\" | \\"bar\\"): number => {
      return store.commit(mutateG, val);
    };
    const mutateH = (): number => {
      return store.commit(\\"mutateH\\");
    };
    const getA = computed(() => {
      return store.getters.getA;
    });
    const getB = computed<number>(() => {
      return store.getters.getB;
    });
    const getTheC = computed(() => {
      return store.getters[\\"namespace/getC\\"];
    });
    const getTheG = computed<number>(() => {
      return store.getters.cGetG;
    });
    const stateA = computed(() => {
      return store.state.stateA;
    });
    const stateB = computed<number>(() => {
      return store.state.stateB;
    });
    const getStateC = computed<Foo>(() => {
      return store.state[\\"namespace/StateC\\"];
    });
    const stateD = computed<boolean>(() => {
      return store.state[ns.keys.stateD];
    });
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myDiv =
      ref<HTMLDivElement>(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const mySpan =
      ref(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myComponent =
      ref<MyComponent>(null);
    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.'); // foo
    onBeforeUnmount(() => {
      context.emit(\\"tearing down\\");
    });
    onUnmounted(() => {
      console.log(\\"destroyed\\");
    });
    const click = (c: string) => {
      context.emit(\\"click\\", c);
    };
    const refAccess = () => {
      const foo = { myDiv: true };
      foo.myDiv = false; // should not transform
      myDiv.value!.focus();
      mySpan.value!.innerText = \\"foo\\";
      myComponent.value!.vm.doSomething();
    };
    const value = computed({
      get: () => {
        return props.checked;
      },
      set: (value: boolean) => {
        context.emit(\\"change\\", value);
      },
    });
    watch(
      props.checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal);
        console.log(anotherComponent.value!);
      },
      { deep: true, immediate: true }
    );
    watch(
      msg.value,
      (val: string, newVal: string) => {
        console.log(val, newVal);
      },
      {}
    );
    onMounted(() => {
      click(\\"oao\\");
    });
    const removeItem = () => {
      if (props.foo > 10) {
        context.emit(\\"remove-item\\", 10);
      } else {
        context.emit(\\"remove-item\\", props.foo);
      }
    };
    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value);
    };
    removeItem();
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      actA,
      actD,
      actE,
      actuhGee,
      actH,
      mutateA,
      mutateB,
      mutateE,
      mGee,
      mutateH,
      getA,
      getB,
      getTheC,
      getTheG,
      stateA,
      stateB,
      getStateC,
      stateD,
      myDiv,
      mySpan,
      myComponent,
      msg,
      onBeforeUnmount,
      onUnmounted,
      click,
      refAccess,
      value,
      checked,
      msg,
      onMounted,
      removeItem,
      hello,
    };
  },
});
"
`;

exports[`testTSFile compatible and ts config file: stderr 1`] = `""`;

exports[`testTSFile compatible and ts config file: stdout 1`] = `
"import {
  computed,
  PropType,
  reactive,
  toRefs,
  ref,
  onBeforeUnmount,
  onUnmounted,
  watch,
  onMounted,
  defineComponent,
} from \\"@vue/composition-api\\";
import { useStore } from \\"vuex\\";
/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: \\"checked\\",
    event: \\"change\\",
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Boolean, default: false },
    bar: { type: Number as PropType<number>, default: 1 },
    foobar: { type: Object as PropType<CustomType> },
    noType: {},
    stringType: { type: String },
    booleanType: { type: Boolean },
    numberType: { type: Number },
    stringLiteralType: { type: String as PropType<\\"foo\\"> },
    numericLiteral: { type: Number as PropType<2> },
    booleanLiteral: { type: Boolean as PropType<true> },
    /* TODO: Too complex to determine primitive type: true | \\"false\\" */ complexType:
      {},
    /* TODO: Too complex to determine primitive type: string[] */ complexType2:
      {},
  },
  name: \\"oao\\",
  setup(props, ctx) {
    const store = useStore();
    /**
     * My count
     */
    const count = computed(() => {
      return /* TODO: Check this convert result, it can work well in 80% of cases.*/ ctx
        .root.$store.state.count;
    });
    const { a } = (() => {
      const a = \\"pa\\";
      return toRefs(
        reactive({
          a: a,
        })
      );
    })();
    const anotherComponent = ref<HTMLElement>(null);
    /* TODO: check function dispatch call signature is correct*/ const actA =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"actA\\", arg);
      };
    /* TODO: check function dispatch call signature is correct*/ const actD =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"namespace/actD\\", arg);
      };
    const actE = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"namespace/actE\\", val);
    };
    const actuhGee = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(actG, val);
    };
    const actH = async (): Promise<number> => {
      return await store.dispatch(\\"actH\\");
    };
    /* TODO: check function dispatch call signature is correct*/ const mutateA =
      (arg: unknown): unknown => {
        return store.commit(\\"mutateA\\", arg);
      };
    /* TODO: check function dispatch call signature is correct*/ const mutateB =
      (arg: unknown): unknown => {
        return store.commit(\\"namespace/mutateB\\", arg);
      };
    const mutateE = (val: \\"foo\\" | \\"bar\\"): number => {
      return store.commit(\\"namespace/mutateE\\", val);
    };
    const mGee = (val: \\"foo\\" | \\"bar\\"): number => {
      return store.commit(mutateG, val);
    };
    const mutateH = (): number => {
      return store.commit(\\"mutateH\\");
    };
    const getA = computed(() => {
      return store.getters.getA;
    });
    const getB = computed<number>(() => {
      return store.getters.getB;
    });
    const getTheC = computed(() => {
      return store.getters[\\"namespace/getC\\"];
    });
    const getTheG = computed<number>(() => {
      return store.getters.cGetG;
    });
    const stateA = computed(() => {
      return store.state.stateA;
    });
    const stateB = computed<number>(() => {
      return store.state.stateB;
    });
    const getStateC = computed<Foo>(() => {
      return store.state[\\"namespace/StateC\\"];
    });
    const stateD = computed<boolean>(() => {
      return store.state[ns.keys.stateD];
    });
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myDiv =
      ref<HTMLDivElement>(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const mySpan =
      ref(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myComponent =
      ref<MyComponent>(null);
    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.'); // foo
    onBeforeUnmount(() => {
      ctx.emit(\\"tearing down\\");
    });
    onUnmounted(() => {
      console.log(\\"destroyed\\");
    });
    const click = (c: string) => {
      context.emit(\\"click\\", c);
    };
    const refAccess = () => {
      const foo = { myDiv: true };
      foo.myDiv = false; // should not transform
      myDiv.value!.focus();
      mySpan.value!.innerText = \\"foo\\";
      myComponent.value!.vm.doSomething();
    };
    const value = computed({
      get: () => {
        return props.checked;
      },
      set: (value: boolean) => {
        ctx.emit(\\"change\\", value);
      },
    });
    watch(
      props.checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal);
        console.log(anotherComponent.value!);
      },
      { deep: true, immediate: true }
    );
    watch(
      msg.value,
      (val: string, newVal: string) => {
        console.log(val, newVal);
      },
      {}
    );
    onMounted(() => {
      click(\\"oao\\");
    });
    const removeItem = () => {
      if (props.foo > 10) {
        context.emit(\\"remove-item\\", 10);
      } else {
        context.emit(\\"remove-item\\", props.foo);
      }
    };
    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value);
    };
    removeItem();
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      actA,
      actD,
      actE,
      actuhGee,
      actH,
      mutateA,
      mutateB,
      mutateE,
      mGee,
      mutateH,
      getA,
      getB,
      getTheC,
      getTheG,
      stateA,
      stateB,
      getStateC,
      stateD,
      myDiv,
      mySpan,
      myComponent,
      msg,
      onBeforeUnmount,
      onUnmounted,
      click,
      refAccess,
      value,
      checked,
      msg,
      onMounted,
      removeItem,
      hello,
    };
  },
});

"
`;

exports[`testTSFile no compatible 1`] = `
"import {
  defineComponent,
  computed,
  PropType,
  reactive,
  toRefs,
  ref,
  onBeforeUnmount,
  onUnmounted,
  watch,
  onMounted,
} from \\"vue\\";
import { useStore } from \\"vuex\\";
/**
 * My basic tag
 */
export default defineComponent({
  model: {
    prop: \\"checked\\",
    event: \\"change\\",
  },
  props: {
    bar: String,
    checked: { type: Boolean },
    /**
     * My foo
     */
    foo: { type: Boolean, default: false },
    bar: { type: Number as PropType<number>, default: 1 },
    foobar: { type: Object as PropType<CustomType> },
    noType: {},
    stringType: { type: String },
    booleanType: { type: Boolean },
    numberType: { type: Number },
    stringLiteralType: { type: String as PropType<\\"foo\\"> },
    numericLiteral: { type: Number as PropType<2> },
    booleanLiteral: { type: Boolean as PropType<true> },
    /* TODO: Too complex to determine primitive type: true | \\"false\\" */ complexType:
      {},
    /* TODO: Too complex to determine primitive type: string[] */ complexType2:
      {},
  },
  name: \\"oao\\",
  setup(props, context) {
    const store = useStore();
    /**
     * My count
     */
    const count = computed(() => {
      return /* TODO: Check this convert result, it can work well in 80% of cases.*/ context
        .root.$store.state.count;
    });
    const { a } = (() => {
      const a = \\"pa\\";
      return toRefs(
        reactive({
          a: a,
        })
      );
    })();
    const anotherComponent = ref<HTMLElement>(null);
    /* TODO: check function dispatch call signature is correct*/ const actA =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"actA\\", arg);
      };
    /* TODO: check function dispatch call signature is correct*/ const actD =
      async (arg: unknown): Promise<unknown> => {
        return await store.dispatch(\\"namespace/actD\\", arg);
      };
    const actE = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(\\"namespace/actE\\", val);
    };
    const actuhGee = async (val: \\"foo\\" | \\"bar\\"): Promise<number> => {
      return await store.dispatch(actG, val);
    };
    const actH = async (): Promise<number> => {
      return await store.dispatch(\\"actH\\");
    };
    /* TODO: check function dispatch call signature is correct*/ const mutateA =
      (arg: unknown): unknown => {
        return store.commit(\\"mutateA\\", arg);
      };
    /* TODO: check function dispatch call signature is correct*/ const mutateB =
      (arg: unknown): unknown => {
        return store.commit(\\"namespace/mutateB\\", arg);
      };
    const mutateE = (val: \\"foo\\" | \\"bar\\"): number => {
      return store.commit(\\"namespace/mutateE\\", val);
    };
    const mGee = (val: \\"foo\\" | \\"bar\\"): number => {
      return store.commit(mutateG, val);
    };
    const mutateH = (): number => {
      return store.commit(\\"mutateH\\");
    };
    const getA = computed(() => {
      return store.getters.getA;
    });
    const getB = computed<number>(() => {
      return store.getters.getB;
    });
    const getTheC = computed(() => {
      return store.getters[\\"namespace/getC\\"];
    });
    const getTheG = computed<number>(() => {
      return store.getters.cGetG;
    });
    const stateA = computed(() => {
      return store.state.stateA;
    });
    const stateB = computed<number>(() => {
      return store.state.stateB;
    });
    const getStateC = computed<Foo>(() => {
      return store.state[\\"namespace/StateC\\"];
    });
    const stateD = computed<boolean>(() => {
      return store.state[ns.keys.stateD];
    });
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myDiv =
      ref<HTMLDivElement>(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const mySpan =
      ref(null);
    /* TODO: Check for potential naming collisions from $refs conversion*/ const myComponent =
      ref<MyComponent>(null);
    /**
     * My msg
     */
    const msg = ref('Vetur means \\"Winter\\" in icelandic.'); // foo
    onBeforeUnmount(() => {
      context.emit(\\"tearing down\\");
    });
    onUnmounted(() => {
      console.log(\\"destroyed\\");
    });
    const click = (c: string) => {
      context.emit(\\"click\\", c);
    };
    const refAccess = () => {
      const foo = { myDiv: true };
      foo.myDiv = false; // should not transform
      myDiv.value!.focus();
      mySpan.value!.innerText = \\"foo\\";
      myComponent.value!.vm.doSomething();
    };
    const value = computed({
      get: () => {
        return props.checked;
      },
      set: (value: boolean) => {
        context.emit(\\"change\\", value);
      },
    });
    watch(
      props.checked,
      (val: boolean, newVal: boolean) => {
        console.log(val, newVal);
        console.log(anotherComponent.value!);
      },
      { deep: true, immediate: true }
    );
    watch(
      msg.value,
      (val: string, newVal: string) => {
        console.log(val, newVal);
      },
      {}
    );
    onMounted(() => {
      click(\\"oao\\");
    });
    const removeItem = () => {
      if (props.foo > 10) {
        context.emit(\\"remove-item\\", 10);
      } else {
        context.emit(\\"remove-item\\", props.foo);
      }
    };
    /**
     * My greeting
     */
    const hello = () => {
      console.log(msg.value);
    };
    removeItem();
    // TODO: Please remove unused return variable
    return {
      count,
      a,
      anotherComponent,
      actA,
      actD,
      actE,
      actuhGee,
      actH,
      mutateA,
      mutateB,
      mutateE,
      mGee,
      mutateH,
      getA,
      getB,
      getTheC,
      getTheG,
      stateA,
      stateB,
      getStateC,
      stateD,
      myDiv,
      mySpan,
      myComponent,
      msg,
      onBeforeUnmount,
      onUnmounted,
      click,
      refAccess,
      value,
      checked,
      msg,
      onMounted,
      removeItem,
      hello,
    };
  },
});
"
`;
